runtime/src/main/java/org/corfudb/runtime/collections/SMRMultiIndex.java

```diff
@@ -0,0 +1,374 @@
+        }
+    }
+
+    @Mutator(name = "putAll")
+    @Override
+    public void putAll(Map<? extends K, ? extends V> m) {
+
+        throw new RuntimeException("###########Operation Not Implemented.##########");
+    }
+
+
+    /** Get a value by it's row key.
+     *
+     * @param key    The row key to retrieve.
+     * @return          The object at the row key, or
+     *                  NULL, if no object was mapped.
+     */
+    @Accessor
+    @Nullable
+    public V getByRowIndex(K key) {
+        return mainMap.get(key);
+    }
+
+    /** Get a value by a given named index.
+     *
+     * @param indexName  The index of the index function from the list
+     *                  passed in at construction time.
+     * @param indexKey The index Key to use.
+     * @return          A set of values which match the given index criteria.
+     */
+    @Accessor
+    public Collection getByNamedIndex(String indexName, I indexKey, ProjectionFunction<K,V,I, ?> projectionFunction) {
+        Map<I,Map<K,V>> indexMap = indexMaps.get(indexName);
+        if (indexMap == null)
+            return Collections.EMPTY_SET;
+        Map<K,V> index = indexMap.get(indexKey);
+        if (index == null)
+            return Collections.EMPTY_SET;
+        List result = new ArrayList();
+        for (Map.Entry<K,V> entry : index.entrySet()) {
+            result.add(projectionFunction.generateProjection(entry.getKey(), entry.getValue(), indexKey));
+        }
+        return result;
+    }
+
+    /** Get a value by a given named index.
+     *
+     * @param indexName  The index of the index function from the list
+     *                  passed in at construction time.
+     * @param indexKey The index Key to use.
+     * @return          A set of values which match the given index criteria.
+     */
+    @Accessor
+    public Collection<Map.Entry<K,V>> getByNamedIndexUnsafe(String indexName, I indexKey) {
+        Map<I,Map<K,V>> indexMap = indexMaps.get(indexName);
+        if (indexMap == null)
+            return Collections.EMPTY_SET;
+        Map<K,V> index = indexMap.get(indexKey);
+        if (index == null)
+            return Collections.EMPTY_SET;
+        return index.entrySet();
+    }
+
+    /** Get the size of the multi-index.
+     * @return  The number of entries mapped (rows).
+     */
+    @Accessor
+    public int size() {
+        return mainMap.size();
+    }
+
+    @Accessor
+    @Override
+    public boolean isEmpty() {
+        return mainMap.isEmpty();
+    }
+
+    @Accessor
+    @Override
+    public boolean containsKey(Object key) {
+        return mainMap.containsKey(key);
+    }
+
+    @Accessor
+    @Override
+    public boolean containsValue(Object value) {
+        return mainMap.containsValue(value);
+    }
+
+
+
+    /** Clear the multi-index.
+     */
+    @Mutator(reset = true, name = "clear")
+    public void clear() {
+        throw new RuntimeException("###########Operation Not Implemented.##########");
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Define and throw a dedicated exception instead of using a generic one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00112)